libs/dyn/pattern_string.go

```diff
@@ -0,0 +1,102 @@
+package dyn
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// MustPatternFromString is like NewPatternFromString but panics on error.
+func MustPatternFromString(input string) Pattern {
+	p, err := NewPatternFromString(input)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// NewPatternFromString parses a pattern from a string.
+//
+// The string must be a sequence of keys and indices separated by dots.
+// Indices must be enclosed in square brackets.
+// The string may include a leading dot.
+// The wildcard character '*' can be used to match any key or index.
+//
+// Examples:
+//   - foo.bar
+//   - foo[1].bar
+//   - foo.*.bar
+//   - foo[*].bar
```

Comment:
This can allow `foo[].bar` for the index wildcard instead -- a bit clearer than the `*` IMO.

Reply:
When I see `*` I read it as for-each and it works for both keys and indices. `foo[]` does no scream "for each" at me.